### PR TITLE
[#1281] Database is locked

### LINKED ISSUE #1281: Database is locked
SQLite.swift package crashes in failableNext() call. This happens occasionally when transaction history screen is calling allTransactions of the SDKSynchronizer. I suspect a race condition/concurrency issue because the SDKSynchonizer is syncing and at that time it's also supposed to fetch the transactions while probably already accessing the DB.

```swift
extension FailableIterator {
    public func next() -> Element? {
        // swiftlint:disable:next force_try
        try! failableNext()
    }
}
```

### PR DESCRIPTION
**Pull request by LukasKorba**
Closes #1281 

- Execution of any query in TransactionSQLDAO was not thread safe. Since all queries goes through execute() method, this is now using a lock

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

**Review by daira — COMMENTED** [review]
Does the lock also need to be taken for `connection().scalar(...)`?

**Comment by LukasKorba** [comment]
> utACK. One minor question; would it be any safer to use `defer` to perform the `globalDBLock.unlock()` calls around the `zcashlc_*` calls (maybe with some smaller scope for the defer)? Or are those already guaranteed not to throw?

Guaranteed not to throw, everything that is preceded with `try` used defer. My goal was to unlock the lock as soon as possible.

File: Sources/ZcashLightClientKit/DAO/TransactionDao.swift [hunk 1]
```diff
@@ -41,6 +41,9 @@ class TransactionSQLDAO: TransactionRepository {
     
     func countAll() async throws -> Int {
         do {
+            globalDBLock.lock()
+            defer { globalDBLock.unlock() }
```
Comment (daira) [inline Sources/ZcashLightClientKit/DAO/TransactionDao.swift:45]: Another option would be to have a `scalar_locked` helped that just takes the argument of `scalar` as a parameter. Non-blocking.

**Review by daira — CHANGES_REQUESTED** [review]
connection.run and connection.transaction need the lock held.

File: Sources/ZcashLightClientKit/DAO/TransactionDao.swift (hunk 1 above)
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/DAO/TransactionDao.swift]: scalarLocked helpers implemented, refactored

File: Sources/ZcashLightClientKit/DAO/UnspentTransactionOutputDao.swift [hunk 2]
```diff
@@ -118,6 +121,9 @@ class UnspentTransactionOutputSQLDAO: UnspentTransactionOutputRepository {
     /// - Throws: `unspentTransactionOutputDAOStore` if sqlite query fails.
     func store(utxos: [UnspentTransactionOutputEntity]) async throws {
         do {
+            globalDBLock.lock()
+            defer { globalDBLock.unlock() }
+
             let db = try dbProvider.connection()
```
Comment (nuttycom) [inline Sources/ZcashLightClientKit/DAO/UnspentTransactionOutputDao.swift:127]:
```suggestion
```

Comment (nuttycom) [reply Sources/ZcashLightClientKit/DAO/UnspentTransactionOutputDao.swift:127]: It doesn't look like that `db` reference is used?

Comment (nuttycom) [reply Sources/ZcashLightClientKit/DAO/UnspentTransactionOutputDao.swift:127]: Or, no, this is weird, should this be `it.run` on line 130?

File: Sources/ZcashLightClientKit/DAO/UnspentTransactionOutputDao.swift [hunk 3]
```diff
@@ -118,6 +121,9 @@ class UnspentTransactionOutputSQLDAO: UnspentTransactionOutputRepository {
...
             try dbProvider.connection().transaction {
```
Comment (nuttycom) [inline Sources/ZcashLightClientKit/DAO/UnspentTransactionOutputDao.swift:128]:
```suggestion
            try db.transaction {
```

**Review by nuttycom — CHANGES_REQUESTED** [review]
Additional changes will be needed in `AccountEntity`, looking for elsewhere where we also have to consider this problem.

**Review by nuttycom — APPROVED** [review]
utACK; I have grepped the codebase and checked that the lock is used everywhere that the `connection()` is used.

**Review by daira — APPROVED** [review]
utACK

**Pull request merged**